File: app/src/main/java/ch/bailu/aat/gpx/GpxListIterator.java
package ch.bailu.aat.gpx;

import ch.bailu.aat.gpx.linked_list.Node;
import ch.bailu.aat.gpx.segmented_list.SegmentNode;

public class GpxListIterator {

    private class PointPrimerNode extends GpxPointFirstNode {
        public PointPrimerNode() {
            super(GpxPoint.NULL, GpxAttributesStatic.NULL_ATTRIBUTES);
        }
        
        @Override
        public Node getNext() {
            return track.getPointList().getFirst();
        }
    }

    
    private class SegmentPrimerNode extends GpxSegmentNode {
        public SegmentPrimerNode() {
            super(new PointPrimerNode());
        }
        
        @Override
        public Node getNext() {
            return track.getSegmentList().getFirst();
        }
    }
    
    private final GpxList track;

    private Node point = new PointPrimerNode();
    private Node segment = new SegmentPrimerNode();
    
    private int inSegmentIndex=-1;
    private int inTrackIndex=-1;
    
    public GpxListIterator(GpxList t) {
        track=t;
    }

    
    public boolean nextPoint() {
        if (setPoint(point.getNext())) {
            inSegmentIndex++;
            inTrackIndex++;
            
            if (inSegmentIndex == ((SegmentNode)segment).getSegmentSize()) 
                return nextSegment();
            
            return true;
        }
        return false;
    }
    
    
    private boolean setPoint(Node n) {
        if (n == null) return false;
        point = n;
        return true;
    }
    
    
    private boolean nextSegment() {
        if (setSegment(segment.getNext())) {
            inSegmentIndex=0;
            return true;
        }
        return false;
    }

    
    private boolean setSegment(Node n) {
        if (n == null) return false;
        segment = n;
        return true;
    }
    
    public GpxPointNode getPoint() {
        return (GpxPointNode)point;
    }


    public boolean isFirstInTrack() {
        return inTrackIndex==0;
    }
    
    public boolean isFirstInSegment() {
        return inSegmentIndex==0;
    }
}

File: aat-lib/src/main/java/ch/bailu/aat_lib/gpx/GpxListIterator.kt
package ch.bailu.aat_lib.gpx

import ch.bailu.aat_lib.gpx.attributes.GpxAttributesNull
import ch.bailu.aat_lib.gpx.linked_list.Node
import ch.bailu.aat_lib.gpx.segmented_list.SegmentNode

/** Forward iterator over a {@link GpxList}, tracking segment boundaries and point indices. */
class GpxListIterator(private val track: GpxList) {
    private inner class PointPrimerNode : GpxPointFirstNode(GpxPoint.NULL, GpxAttributesNull.NULL) {
        override var next: Node?
            get() = track.pointList.first
            set(_) {}
    }

    private inner class SegmentPrimerNode : GpxSegmentNode(PointPrimerNode()) {
        override var next: Node?
            get() = track.segmentList.first
            set(_) {}
    }

    private var point: GpxPointNode = PointPrimerNode()
    private var segment: GpxSegmentNode = SegmentPrimerNode()

    private var inSegmentIndex = -1
    private var inTrackIndex = -1


    fun nextPoint(): Boolean {
        if (setPoint(point.next)) {
            inSegmentIndex++
            inTrackIndex++

            if (inSegmentIndex == (segment as SegmentNode).segmentSize) return nextSegment()
            return true
        }
        return false
    }

    private fun setPoint(node: Node?): Boolean {
        if (node is GpxPointNode) {
            point = node
            return true
        }
        return false
    }

    private fun nextSegment(): Boolean {
        if (setSegment(segment.next)) {
            inSegmentIndex = 0
            return true
        }
        return false
    }

    private fun setSegment(node: Node?): Boolean {
        if (node is GpxSegmentNode) {
            segment = node
            return true
        }
        return false
    }

    fun getPoint(): GpxPointNode {
        return point
    }

    val isFirstInTrack: Boolean
        get() = inTrackIndex == 0

    val isFirstInSegment: Boolean
        get() = inSegmentIndex == 0
}
